trin-core/src/portalnet/overlay.rs

```diff
@@ -192,12 +201,160 @@ where
     }
 
     /// Process overlay uTP listener event
-    pub fn process_utp_event(&self, event: UtpListenerEvent) {
-        // TODO: Handle overlay uTP events
-        debug!(
-            "Got overlay uTP event: {event:?}, protocol id: {:?}",
-            self.protocol
-        );
+    pub fn process_utp_event(&self, event: UtpListenerEvent) -> anyhow::Result<()> {
+        match event {
+            UtpListenerEvent::ClosedStream(utp_payload, _, stream_id) => {
+                // Handle all closed uTP streams, currently we process only AcceptStream here.
+                // FindContent payload is processed explicitly when we send FindContent request.
+                // We don't expect to process a payload for ContentStream and OfferStream.
+                match stream_id {
+                    UtpStreamId::AcceptStream(content_keys) => {
+                        self.process_accept_utp_payload(content_keys, utp_payload)?
+                    }
+                    UtpStreamId::ContentStream(_) => {}
+                    UtpStreamId::FindContentStream => {}
+                    UtpStreamId::OfferStream => {}
+                }
+            }
+            UtpListenerEvent::ResetStream(..) => {}
+        }
+        Ok(())
+    }
+
+    /// Process accepted uTP payload of the OFFER?ACCEPT stream
+    fn process_accept_utp_payload(
+        &self,
+        content_keys: Vec<RawContentKey>,
+        payload: UtpPayload,
+    ) -> anyhow::Result<()> {
+        // Process accepted uTP payload as SZZ decoded Container[payloads: List[ByteList, max_length=64]]
+        let content_values = ContentPayloadList::from_ssz_bytes(&payload)
+            .map_err(|err| anyhow!("Unable to decode uTP content payload: {err:?}"))?;
+        // Max content value length should be 64
+        if content_values.len() > 64 {
+            return Err(anyhow!("Received content values with len > 64."));
+        }
+        // Accepted content keys len should match content value len
+        if content_keys.len() != content_values.len() {
+            return Err(anyhow!(
+                "Content keys len doesn't match content values len."
+            ));
+        }
+
+        // TODO: Verify overlay content data with an Oracle
```

Comment:
This should just require an update from the current `phantom_validator` field to be a handle to the actual `HeaderOracle`, which we can then use to validate the content.